config/custom.go

```diff
@@ -95,58 +99,227 @@ func (slc *StringSlice) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	// zero means not assigned
+	LbWeight uint32
+}
+
+func (u *WeightedURL) Validate() error {
+	if u.URL.Hostname() == "" {
+		return errHostnameMustBeSpecified
+	}
+	if u.URL.Scheme == "" {
+		return errSchemeMustBeSpecified
+	}
+	return nil
+}
+
+// ParseWeightedURL parses url that has an optional weight appended to it
+func ParseWeightedURL(dst string) (*WeightedURL, error) {
+	to, w, err := weightedString(dst)
+	if err != nil {
+		return nil, err
+	}
+
+	u, err := url.Parse(to)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", to, err)
+	}
+
+	if u.Hostname() == "" {
+		return nil, errHostnameMustBeSpecified
+	}
+
+	return &WeightedURL{*u, w}, nil
+}
+
+func (u *WeightedURL) String() string {
+	str := u.URL.String()
+	if u.LbWeight == 0 {
+		return str
+	}
+	return fmt.Sprintf("{url=%s, weight=%d}", str, u.LbWeight)
+}
+
+type WeightedURLs []WeightedURL
+
+// ParseWeightedUrls parses
+func ParseWeightedUrls(urls ...string) (WeightedURLs, error) {
+	out := make([]WeightedURL, 0, len(urls))
+
+	for _, dst := range urls {
+		u, err := ParseWeightedURL(dst)
+		if err != nil {
+			return nil, err
+		}
+		out = append(out, *u)
+	}
+
+	if _, err := WeightedURLs(out).Validate(); err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// HasWeight indicates if url group has weights assigned
+type HasWeight bool
+
+// Validate checks that URLs are valid, and either all or none have weights assigned
+func (urls WeightedURLs) Validate() (HasWeight, error) {
+	noWeight := false
+	hasWeight := false
+
+	for i := range urls {
+		if err := urls[i].Validate(); err != nil {
+			return false, fmt.Errorf("%s: %w", urls[i].String(), err)
 		}
+		if urls[i].LbWeight == 0 {
+			noWeight = true
+		} else {
+			hasWeight = true
+		}
+	}
 
-		m, ok := data.(map[string]interface{})
-		if !ok {
+	if noWeight == hasWeight {
+		return false, errEndpointWeightsSpec
+	}
+
+	if noWeight {
+		return HasWeight(false), nil
+	}
+	return HasWeight(true), nil
+}
+
+// Flatten converts weighted url array into indidual arrays of urls and weights
+func (urls WeightedURLs) Flatten() ([]string, []uint32, error) {
+	hasWeight, err := urls.Validate()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	str := make([]string, 0, len(urls))
+	wghts := make([]uint32, 0, len(urls))
+
+	for i := range urls {
+		str = append(str, urls[i].URL.String())
+		wghts = append(wghts, urls[i].LbWeight)
+	}
+
+	if !hasWeight {
+		return str, nil, nil
+	}
+	return str, wghts, nil
+}
+
+func DecodePolicyHookFunc() mapstructure.DecodeHookFunc {
+	return func(f, t reflect.Type, data interface{}) (interface{}, error) {
+		if t != reflect.TypeOf(Policy{}) {
 			return data, nil
 		}
 
-		ps, ok := m[policyKey].([]interface{})
+		log.Printf("POLICY HOOK %#v\n", data)
+
+		mp, err := serializable(data)
+		if err != nil {
+			return nil, err
+		}
+		ms, ok := mp.(map[string]interface{})
 		if !ok {
-			return data, nil
+			return nil, errKeysMustBeStrings
```

Comment:
Have you tested this? When I originally wrote this function the type was a `map[interface{}]interface{}` even though all the keys were strings.

Reply:
yes; the call to `serializable` above does `map[interface{}]interface{} => map[string]interface{}` conversion; thus it should be an unreachable error condition, but exists just in case to assert type conversion.  

```golang
		mp, err := serializable(data)
		if err != nil {
			return nil, err
		}

```